### ISSUE TITLE
Saving state_dict of optimizer

### ISSUE THREAD

**Issue by danpovey** [k2-fsa/snowfall#110]
Guys, I notice the optimizer code e.g. for Noam, has load_state_dict and save_state_dict, but these appear not to be used and its step gets set to zero if you try to train from an intermediate epoch.  This means that with the attention models you cannot restart training and get the same results.

Not sure what the right fix for this is.

**Comment by pzelasko** [comment]
We'll need to add it as an argument to save/load checkpoint functions and store the state dict in checkpoints. IIRC the learning rate schedulers in pytorch have the same mechanism. I can do it but not sure when; if somebody's quicker then feel free to have a go.

**Comment by zhu-han** [comment]
Fixed in https://github.com/k2-fsa/snowfall/pull/111, like @pzelasko said, save optimizer and lr_scheduler every epoch and load them if training from an intermediate epoch.
